### PR TITLE
Blog - Go with the Gin Framework

### PR DESCRIPTION
**Pull request by msali123**
Done with all checks.

### CODE REVIEW

**Comment by msali123** [comment]
Preview URL:https://63d2ae6ddf9742088ba2eae3--ephemeral-narwhal-3ac3b2.netlify.app

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+
+<div style="height= 50%">
+![Starting]({{site.images}}{{page.slug}}/start.png)\
+</div>
+
+Once you've set up your Go workspace, install the `gin` package in your working directory using this command.
+
+~~~{.go caption=">_"}
+go get github.com/gin-gonic/gin
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:48]:
```suggestion
~~~{.bash caption=">_"}
go get github.com/gin-gonic/gin
~~~
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+You'll also need to install the `gorm` package and the `gorm` [sqlite](/blog/golang-sqlite) driver for connecting to the SQLite database.
+
+Run these commands in your working directory to install the [packages](/blog/setup-typescript-monorepo).
+
+~~~{.go caption=">_"}
+go get gorm.io/gorm
+go get gorm.io/driver/sqlite
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:57]:
```suggestion
~~~{.bash caption=">_"}
go get gorm.io/gorm
go get gorm.io/driver/sqlite
~~~
```

File: blog/_posts/2023-02-27-golang-gin-framework.md (hunk 1 above)
Comment (balapriyac) [reply blog/_posts/2023-02-27-golang-gin-framework.md:48]: We run the `go get` command at the terminal.

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+These are the imports you'll need for this tutorial.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:59]:
```suggestion
These are the imports you'll need for this tutorial:
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+Gin is extendable, and the package provides built-in rendering support for HTML, [JSON](/blog/convert-to-from-json), and XML rendering with documentation support, among many other features.
+
+This tutorial will walk you through developing web applications in Go with the Gin framework. You'll learn how to use Gin by building a CRUD API. You'll use the Gin framework for routing, JSON parsing, request-response-related operations, and the GORM package for the database (SQLite) auto migrations and operations.
+
+Note: You can find the complete code for this tutorial on this [GitHub Gist](https://gist.github.com/Goodnessuc/e4fdc78b04965e991efc440f50705ece)
+
+## Prerequisites
+
+You'll need to meet a few prerequisites to understand and follow this hands-on tutorial.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:33]:
```suggestion
You'll need to meet a few prerequisites to understand and follow this hands-on tutorial:
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+<div class="wide">
+
+![Untitled]({{site.images}}{{page.slug}}/Untitled.png)\
+</div>
+
+Gin is extendable, and the package provides built-in rendering support for HTML, [JSON](/blog/convert-to-from-json), and XML rendering with documentation support, among many other features.
+
+This tutorial will walk you through developing web applications in Go with the Gin framework. You'll learn how to use Gin by building a CRUD API. You'll use the Gin framework for routing, JSON parsing, request-response-related operations, and the GORM package for the database (SQLite) auto migrations and operations.
+
+Note: You can find the complete code for this tutorial on this [GitHub Gist](https://gist.github.com/Goodnessuc/e4fdc78b04965e991efc440f50705ece)
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:29]:
```suggestion
Note: You can find the complete code for this tutorial on this [GitHub Gist](https://gist.github.com/Goodnessuc/e4fdc78b04965e991efc440f50705ece).
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+
+Once you've set up your Go workspace, install the `gin` package in your working directory using this command.
+
+~~~{.go caption=">_"}
+go get github.com/gin-gonic/gin
+~~~
+
+You'll also need to install the `gorm` package and the `gorm` [sqlite](/blog/golang-sqlite) driver for connecting to the SQLite database.
+
+Run these commands in your working directory to install the [packages](/blog/setup-typescript-monorepo).
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:52]:
```suggestion
Run these commands in your working directory to install the packages:
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+)
+~~~
+
+You'll use the `log` package for logging-related operations and the `http` package for starting a server and other operations.
+
+## Setting Up GORM and SQLite for Persistence
+
+GORM uses structs for the database model. You can declare the struct with constraint tags and set up auto migrations for the struct.
+
+Here's an example `company` struct with `gorm` and `json` tags.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:77]:
```suggestion
Here's an example `company` struct with `gorm` and `json` tags:
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+}
+~~~
+
+After creating the router instance, you can use the `GET`, `POST`, `PUT`, and `DELETE` methods to define routes and their respective handler functions.
+
+The `Run` method of your router instance starts a server to run on the specified port. You can now define your handler functions and mount them in the main function.
+
+### The `POST` Request Handler
+
+The `POST` request handler function will accept JSON input from the client for GORM to migrate to the decoded JSON struct database.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:150]:
```suggestion
The `POST` request handler function will accept JSON input from the client for GORM to migrate to the decoded JSON struct database:
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+    if err := db.Create(&newCompany).Error; err != nil {
+        ctx.JSON(http.StatusInternalServerError, \
+        gin.H{"error": "Database Migration Error"})
+    }
+    ctx.JSON(http.StatusOK, company)
+
+}
+~~~
+
+The PostCompany handler function receives POST requests from the server, parses the JSON into the `newCompany` struct, and the Create method of your database instance creates a new row of the struct inputs in the database.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:177]:
```suggestion


```

Comment (balapriyac) [reply blog/_posts/2023-02-27-golang-gin-framework.md:177]: Remove repeated paragraph.

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+
+~~~{.go caption="gin.go"}
+func main() {
+    router := Gin.Default()
+    router.POST("/company", PostCompany)
+    log.Fatal(router.Run(":8080"))
+}
+~~~
+
+Here's a CURL request that tests the `PostCompany` handler function.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:193]:
```suggestion
Here's a CURL request that tests the `PostCompany` handler function:
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+    router := Gin.Default()
+    router.GET("api/v1/:company", GetCompany)
+    router.POST("/company", PostCompany)
+
+    log.Fatal(router.Run(":8080"))
+}
+
+~~~
+
+Here's the CURL request for the `GetCompany` handler function. The `CURL` request sends a request to the `/api/v1/:company` route with the data attached to the URL.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:244]:
```suggestion
Here's the CURL request for the `GetCompany` handler function. The `CURL` request sends a request to the `/api/v1/:company` route with the data attached to the URL:
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+    router := Gin.Default()
+    router.GET("api/v1/:company", GetCompany)
+    router.POST("/company", PostCompany)
+    router.PUT("api/v1/:company", UpdateCompany)
+
+    log.Fatal(router.Run(":8080"))
+}
+~~~
+
+Here's the CURL request that tests the `UpdateCompany` handler function. Insert a company name in the specified field to run the CURL request effectively.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:314]:
```suggestion
Here's the CURL request that tests the `UpdateCompany` handler function. Insert a company name in the specified field to run the CURL request effectively:
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+    router.GET("api/v1/:company", GetCompany)
+    router.POST("/company", PostCompany)
+    router.PUT("api/v1/:company", UpdateCompany)
+    router.DELETE("api/v1/:company", DeleteCompany)
+
+    log.Fatal(router.Run(":8080"))
+}
+~~~
+
+Here's the CURL request for the `DeleteCompany` handler function. Insert a company name in the specified field to run the CURL request effectively.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:380]:
```suggestion
Here's the CURL request for the `DeleteCompany` handler function. Insert a company name in the specified field to run the CURL request effectively:
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+curl -X PUT -H "Content-Type: application/json" \
+-d '{"name": "TestCompany", "created": "2022-01-01", \
+"product": "UpdatedProduct"}' "http://localhost:8080/api/v1/<company_name>"
+
+{"name": "TestCompany", "created": "2022-01-01", \
+"product": "UpdatedProduct "}
+
+~~~
+
+The CURL request sends a PUT request to the `api/v1/:company` endpoint with a JSON payload as the replacement for the update operation.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:326]:
```suggestion
The CURL request sends a PUT request to the `api/v1/:company` endpoint with a [JSON](/blog/convert-to-from-json) payload as the replacement for the update operation.
```

File: blog/_posts/2023-02-27-golang-gin-framework.md [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+
+Here's the CURL request for the `DeleteCompany` handler function. Insert a company name in the specified field to run the CURL request effectively.
+
+~~~{.bash caption=">_"}
+    curl -X DELETE "http://localhost:8080/api/v1/<company_name>"
+
+     {"message": "Company Deleted"% }
+~~~
+
+The CURL request sends a PUT request to the `api/v1/:company` endpoint with a [JSON](/blog/convert-to-from-json) payload as the replacement for the update operation.
```
Comment (balapriyac) [inline blog/_posts/2023-02-27-golang-gin-framework.md:388]:
```suggestion
The CURL request sends a DELETE request to the `api/v1/:company` endpoint to delete the matching row from the database.
```

Comment (balapriyac) [reply blog/_posts/2023-02-27-golang-gin-framework.md:388]: The author has pasted this explanation from _The PUT Request Handler_ section. Rewrote that sentence as this section is on _The DELETE Request Handler_.

**Review by balapriyac — CHANGES_REQUESTED** [review]
Thank you, Sameer! @jalletto I've finished reviewing this PR.

**Comment by msali123** [comment]
@balapriyac done with the requested changes.

**Review by jalletto — APPROVED** [review]

**Pull request merged**